pallets/rewards/src/lib.rs

```diff
@@ -0,0 +1,546 @@
+			);
+
+			//check if user is added in reward list
+			ensure!(<Distributor<T>>::contains_key(reward_id, &user), Error::<T>::UserNotEligible);
+
+			<Distributor<T>>::mutate(reward_id, user.clone(), |user_reward_info| {
+				if let Some(user_reward_info) = user_reward_info {
+					// only unlock reward if current block greater than or equal to the starting
+					// block of reward
+					if let Some(reward_info) = <InitializeRewards<T>>::get(reward_id) {
+						ensure!(
+							reward_info.start_block.saturated_into::<u128>() <=
+								<frame_system::Pallet<T>>::block_number()
+									.saturated_into::<u128>(),
+							Error::<T>::RewardsCannotBeUnlockYet
+						);
+					}
+
+					//check if user already unlocked the rewards
+					ensure!(!user_reward_info.is_initialized, Error::<T>::RewardsAlreadyUnlocked);
+					//transfer funds from pallet account to users account
+					ensure!(
+						Self::transfer_pdex_rewards(
+							&Self::get_pallet_account(),
+							&user,
+							user_reward_info.total_reward_amount
+						)
+						.is_ok(),
+						Error::<T>::TransferFailed
+					);
+
+					//lock funds in users account
+					T::NativeCurrency::set_lock(
+						REWARDS_LOCK_ID,
+						&user,
+						user_reward_info.total_reward_amount.saturated_into(),
+						WithdrawReasons::TRANSFER,
+					);
+					user_reward_info.is_initialized = true;
+					Ok(())
+				} else {
+					//sanity check
+					Err(Error::<T>::UserNotEligible)
+				}
+			})?;
+			Self::deposit_event(Event::UserUnlockedReward { user, reward_id });
+			Ok(())
+		}
+
+		/// The user will use the extrinsic to claim rewards
+		/// origin: The users address which has been mapped to reward id
+		/// id: The reward id
+		#[pallet::weight(10_000)]
+		pub fn claim(origin: OriginFor<T>, reward_id: u32) -> DispatchResult {
+			let user: T::AccountId = ensure_signed(origin)?;
+
+			//check if given id valid or not
+			ensure!(
+				<InitializeRewards<T>>::contains_key(reward_id),
+				Error::<T>::RewardIdNotRegister
+			);
+
+			//check if user is added in reward list
+			ensure!(<Distributor<T>>::contains_key(reward_id, &user), Error::<T>::UserNotEligible);
+
+			<Distributor<T>>::mutate(reward_id, user.clone(), |user_reward_info| {
+				if let Some(reward_info) = <InitializeRewards<T>>::get(reward_id) {
+					if let Some(user_reward_info) = user_reward_info {
+						//check if user has initialize rewards or not
+						ensure!(
+							user_reward_info.is_initialized,
+							Error::<T>::UserHasNotIntializeClaimRewards
+						);
+
+						//ensure that all rewards are not already claimed
+						ensure!(
+							user_reward_info.claim_amount.saturated_into::<u128>() !=
+								user_reward_info.total_reward_amount.saturated_into::<u128>(),
+							Error::<T>::AllRewardsAlreadyClaimed
+						);
+
+						let mut rewards_claimable: u128 = 0_u128.saturated_into();
+
+						//calculate the intial rewards that can be claimed
+						let initial_rewards_claimed = user_reward_info
```

Comment:
Should this only be calculated if the user has not already claimed the rewards? Instead of calculating then checking?

Reply:
it is also used for calculating the factor so it doesn't matter if u calculate it afterward.

Reply:
I like your implementation, that looks great